docs/sxid.1.in
.Dd July 29, 2013
.Dt SXID 1
.Os
.Sh NAME
.Nm sxid
.Nd check for changes in s[ug]id files and directories
.Sh SYNOPSIS
.Nm sxid
.Bk
.Op Fl c, -config Ar file
.Op Fl n, -nomail
.Op Fl k, -spotcheck
.Op Fl l, -listall
.Op Fl h, -help
.Op Fl V, -version
.Ek
.Sh DESCRIPTION
.Nm sXid
checks for changes in suid and sgid files and directories based on its last
check. Logs are stored by default in
.Pa /var/log/sxid.log .
The changes are then emailed to the address specified in the configuration
file. The default location for the config file is
.Pa @TMPCONF@/sxid.conf
but this can be overridden with the
.Fl -config
option and specifying an
alternate
.Ar file .
.Sh OPTIONS
.Bl -tag -width Ds
.It Fl c, -config Ar file
Specifies an alternate configuration
.Ar file .
.It Fl n, -nomail
Sends output to stdout instead of emailing, useful for spot checks.
.It Fl k, -spotcheck
Checks for changes by recursing the current working directory. Log files
will not be rotated and no email sent. All output will go to stdout.
.It Fl l, -listall
Useful when doing
.Fl -spotcheck
or
.Fl -nomail
to list all files that are logged, regardless of changes.
.It Fl h, -help
Display a brief help message.
.It Fl V, -version
Print version and exit.
.El
.Sh OUTPUT
.Pp
The program outputs several different checks concerning the current status of
the suid and sgid files and directories on the system on which it was
run. This is a basic overview of the format.
.Pp
In the add remove section, new files are preceded by a
.Dq + ,
old ones
are preceded by a
.Dq - .
Note that removed does not mean gone from the filesystem, just that it
is no longer sgid or suid.
.Pp
Most of it is pretty easy to understand. On the sections that show
changes in the file's info (uid, gid, modes...) the format is
old->new. So if the old owner was
.Dq mail
and it is now
.Dq root
then it shows it as mail->root.
.Pp
The list of files in the checks is in the following format:
.Bd -literal
/full/path   *user.group   MODE
.Ed
.Pp
MODE is the 4 digit mode, as in 4755.
.Pp
In the changes section, if the line is preceded by an
.Dq i
then that item has changed inodes since the last check (regardless of any
s[ug]id change), if there is an
.Dq m
then the SHA-256 checksum has changed.
.Pp
If a user or group entry is preceded by a
.Dq *
then its execution bit is set (i.e. *root.wheel is suid, root.*wheel
is sgid, *root.*wheel is +s).
.Pp
On the forbidden directories, if
.Va ENFORCE
is enabled an
.Dq r
will precede forbidden items that were successfully \-s'd, and an
.Do
!
.Dc
will show that it was unsuccessfully \-s'd (for what ever reason).
.Sh AUTHOR
.An Ben Collins Aq bcollins@debian.org
.Sh "REPORTING BUGS"
.An Timur Birsh Aq taem@linukz.org
.Sh "SEE ALSO"
.Xr sxid.conf 5


NEWS
sXid 4.20130802
  * New version scheme - V.YYYYMMDD. V is a major version number,
    YYYYMMDD is a date of sXid release.
  * Now sXid uses SHA-256 hash function to track files.
    MD5 is considered cryptographically broken.
    Please see http://www.kb.cert.org/vuls/id/836068.
    SHA-256 implementation is taken from PolarSSL and distributed
    under the terms GNU General Public License Version 2 or later.
    Please see https://polarssl.org.
  * autoconf settings updated to version 2.69.
  * Use automake 1.13.3 to generate Makefiles.
  * --version option added.
  * /media and /sys directories added to EXCLUDE option.
  * Manpages converted to mdoc macro.
  * README converted to Markdown format.
  * Some directories and files reorganisation.
  * License file COPYING added.
  * Minor code improvements.
  * Misc bug fixes.

sXid 4.2

  * Release used resources. Resource leaks found by cppcheck:
    
    [./source/logging.c:374]: (error) Resource leak: log
    [./source/md5sum.c:60]: (error) Resource leak: fp
    [./source/search.c:171]: (error) Resource leak: list
    
    Thanks to George Danchev for pointing this out.

 -- Timur Birsh <taem@linukz.org>  Wed, 14 Sep 2009 21:11:45 +0600

sXid 4.1

  * New maintainer.
  * New homepage - http://linukz.org/sxid.shtml.
  * Fix Debian bug#412853 'sxid reports many false positive md5sums changes'.
    Thanks to Ari Johnson <iamtheari@gmail.com>.
  * Update FSF address in copyright notice.

 -- Timur Birsh <taem@linukz.org>  Wed, 09 Sep 2009 15:18:27 +0600


src/sxid.c
/*
 * sxid - suid, sgid file and directory checking
 * sxid.c - main program routines
 *
 * Copyright © 1999, 2000, 2002 Ben Collins <bcollins@debian.org>
 * Copyright © 2009, 2013 Timur Birsh <taem@linukz.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h> /* access() */
#include <string.h> /* strerror() */
#include <errno.h>
#include <time.h>
#include <getopt.h>

#include "sxid.h"

struct option cmd_options[] =
{
	{ "nomail",    no_argument,	     0, 'n' },
	{ "help",      no_argument,	     0, 'h' },
	{ "version",   no_argument,	     0, 'V' },
	{ "listall",   no_argument,	     0, 'l' },
	{ "spotcheck", no_argument,	     0, 'k' },
	{ "config",    required_argument,    0, 'c' },
	{ NULL,	       0,		     0, 0   }
};

char sxid_version[] = VERSION;

int changed = 0;
int notify_diff = 0;
int nomail = 0;

static
void sxid_usage()
{
	fprintf(stdout, "Usage: sxid [options]\n");
	fprintf(stdout, "-h, --help              This help output\n");
	fprintf(stdout, "-V, --version           Print version and exit\n");
	fprintf(stdout, "-k, --spotcheck         Run this using spotcheck rules "
		"(see man page)\n");
	fprintf(stdout, "-n, --nomail            Send output to stdout instead "
		"of sending email\n");
	fprintf(stdout, "-l, --listall           List all found s[ug]id files "
		"regardless of changes\n");
	fprintf(stdout, "-c, --config <file>     Specifies alternative config "
		"file\n");
}

static
void print_version()
{
	fprintf(stdout, "sXid %s\n", sxid_version);
	fprintf(stdout, "Copyright (C) 1999, 2000, 2002 Ben Collins\n");
	fprintf(stdout, "Copyright (C) 2009, 2013 Timur Birsh\n");
	fprintf(stdout, "This is free software; see the source for copying "
		"conditions.  There is NO\n"
		"warranty; not even for MERCHANTABILITY or FITNESS FOR A "
		"PARTICULAR PURPOSE.\n");
	exit(EXIT_SUCCESS);
}

int main(int argc, char *argv[])
{
	FILE *conffile;
	int opt;
	time_t time_mark;

	umask(077);
	config_options.flags = 0x00;

	while ((opt = getopt_long_only(argc, argv, "nhVlkc:", cmd_options,
						NULL)) != EOF) {
		switch (opt) {
		case 'c':
			confname = optarg;
			break;
		case 'n':
			nomail = 1;
			break;
		case 'l':
			config_options.flags |= FLAG_LALL;
			break;
		case 'k':
			config_options.flags |= FLAG_SPOT;
			break;
		case 'V':
			print_version();
			break;
		case 'h':
			sxid_usage();
			exit(EXIT_SUCCESS);
		default:
			sxid_usage();
			exit(EXIT_FAILURE);
		}
	}
	argc -= optind;
	argv += optind;

	if (access(confname, R_OK) != 0) {
		fprintf(stderr, "E: %s: %s\n", strerror(errno), confname);
		exit(EXIT_FAILURE);
	}

	/* Ok, let's time stamp */
	time_mark = time(NULL);
	if (!strftime(current_time, sizeof(current_time), "%c",
		      localtime(&time_mark)))
		current_time[sizeof(current_time) - 1] = '\0';

	init_conf();
	init_file_entries();
	start_search();
	compare_output();

	if ((changed || (config_options.flags & FLAG_AROT)) &&
	    !(config_options.flags & FLAG_SPOT))
		save_and_rotate();

	if (changed || notify_diff || nomail ||
	    (config_options.flags & (FLAG_ANOT | FLAG_LALL)))
		notify_email();

	fclose(mail_file);
	return 0;
}
